core/src/avm2/object.rs

```diff
@@ -275,17 +334,43 @@ pub trait TObject<'gc>: 'gc + Collect + Debug + Into<Object<'gc>> + Clone + Copy
             {
                 let local_name: Result<AvmString<'gc>, Error> = multiname
                     .local_name()
-                    .ok_or_else(|| "Cannot set undefined property using any name".into());
-                let name = QName::dynamic_name(local_name?);
-                return self.init_property_local(receiver, &name, value, activation);
+                    .ok_or_else(|| "Cannot init undefined property using any name".into());
+                name = Some(QName::dynamic_name(local_name?));
+            } else {
+                return Err(format!(
+                    "Cannot init undefined property {:?}",
+                    multiname.local_name()
+                )
+                .into());
             }
+        }
 
-            return Err(
-                format!("Cannot set undefined property {:?}", multiname.local_name()).into(),
-            );
+        // At this point, name resolution should have completed.
+        let name = name.unwrap();
+
+        // Reject attempts to overwrite lazy-bound methods before they have
+        // been bound.
+        if let Some(class) = self.instance_of() {
+            if class.instance_method(&name)?.is_some() {
+                return Err(format!(
+                    "Cannot overwrite read-only property {:?}",
+                    name.local_name()
+                )
+                .into());
+            }
         }
 
-        self.init_property_local(receiver, &name.unwrap(), value, activation)
+        if let Some(class) = self.as_class_object() {
+            if class.class_method(&name)?.is_some() {
+                return Err(format!(
+                    "Cannot overwrite read-only property {:?}",
+                    name.local_name()
+                )
+                .into());
+            }
+        }
+
```

Comment:
Thoughts for the future: there's not really a difference between a class method and instance method; these are both just traits on some object. For an object instance, the traits are the instance methods, and for the class object, the traits are the class methods, so this could simplify into a single code path.

Similarly, the read-only check would be generalized to looking at the traits, bailing out on all non-slot/setter traits, and this completely replace the Property::READ_ONLY flag (I believe DontEnum is the only flag avmplus actually stores on properties; the other flags are implied by traits).

Reply:
We talked about this on Discord a lot, AFAIK this PR was one of the changes blocking the "big refactor" of moving the instance traits away from instances themselves to the class object; so I think everything checks out here :D